### PR TITLE
Add internal iterator API for accessing relevant chain blocks

### PR DESCRIPTION
**Pull request by yaahc**
## Motivation

The difficulty calculation logic needs the ability to access the last 28 blocks of a block's relevant chain. Before this PR the state internals have had APIs for getting individual blocks but no easy way to abstract over "the last n blocks of a chain starting at this block".

## Solution

This PR adds a `chain` method to `StateService` in `zebra-state`. This function provides an iterator API that abstracts over the `StateService`. It will first attempt to find the given block from the non-finalized state, and if it doesn't exist there it falls back to the finalized state.

The code in this pull request has:
  - [x] Documentation Comments
  - [ ] Unit Tests and Property Tests

## Review

@teor2345

### CODE REVIEW

File: zebra-state/src/service.rs
Comment (teor2345) [inline zebra-state/src/service.rs]: Suggestion:
These functions use any chain, but the rest of the functions in the service use the best chain.
Should we rename them to make that clearer?

```suggestion
    pub fn utxo_any_chain(&self, outpoint: &transparent::OutPoint) -> Option<transparent::Output> {
        self.mem.utxo(outpoint).or_else(|| self.sled.utxo(outpoint))
    }

    /// Return an iterator over the relevant chain of the block identified by
    /// `hash_or_height`.
    #[allow(dead_code)]
    pub fn any_chain(&self, hash_or_height: HashOrHeight) -> Iter<'_> {
```

If we do this rename, we should also rename the memory and state `utxo` methods.

File: zebra-state/src/service.rs
Comment (teor2345) [inline zebra-state/src/service.rs]: I was confused by the state enum and attribute, I thought they were related to the state service.

```suggestion
struct Iter<'a> {
    service: &'a StateService,
    iter_state: IterState,
}

enum IterState {
```

File: zebra-state/src/service.rs
Comment (teor2345) [reply zebra-state/src/service.rs]: Hmm, actually, hashes use any chain, and height uses the best chain.

Let's split the hash and height into different methods?
(I don't need the height method right now, but it seems useful in future.)

File: zebra-state/src/service.rs
Comment (teor2345) [inline zebra-state/src/service.rs]: `block` returns the block in the current best chain, but `chain` needs to return side-chain blocks, if it's given a side-chain hash.

File: zebra-state/src/service.rs
Comment (yaahc) [reply zebra-state/src/service.rs]: I feel like this discussion should be split into a different PR / issue. I had a similar discussion with @hdevalence when I was pairing on the block locator fix where we did some cleanup to these methods to make it uniform across the finalized state, non-finalized state, and service. In that discussion I raised similar concerns about how some of the methods only worked with the best chain and others worked with multiple pieces of data and we decided it was fine. I don't remember precisely what points were brought up though so we should probably split this into a complete and separate discussion about how to structure our internal APIs and get henry's input because I can't recall his points.

File: zebra-state/src/service.rs
Comment (teor2345) [inline zebra-state/src/service.rs]: Can the caller hold the iterator across a chain reorganisation?

If the caller supplies a height on the best chain, this code turns the iterator into hash-based iterator on the best chain at the time the iterator's `next` method is first called.

Instead, let's convert to a hash when the iterator is created, so we choose the best chain at that time. I think those semantics are a bit clearer.

And let's document this potentially surprising behaviour in the `chain` function doc comment.

Comment (yaahc) [reply zebra-state/src/service.rs]: > Can the caller hold the iterator across a chain reorganisation?

No, the caller won't be able to mutate the StateService at all while the iterator exists. Which should also mean that there's no reason to eagerly convert the height to a hash.

Edit: Actually, I think it would still be possible to mutate the finalized state, since sled allows internal mutability, but because we use that as append only and it only has 1 chain I think that is not even theoretically a problem, unless someone else introduced code to intentionally modify the finalized state, which wouldn't ever get through code review.

File: zebra-state/src/service.rs
Comment (teor2345) [inline zebra-state/src/service.rs]: It's a bit confusing for the finalized state function to access the `NonFinalized` enum variant.

Instead, let's have an additional enum variant for newly initialised iterators, before the first call to `next`.

File: zebra-state/src/service.rs
Comment (teor2345) [inline zebra-state/src/service.rs:368]: Let's also implement:
* `Iterator::size_hint`
  * return the height of the hash in the iterator's state (or the height in the iterator's state)
  * https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.size_hint
* `ExactSizeIterator`:
  * the default `len` implementation is good
  * https://doc.rust-lang.org/std/iter/trait.ExactSizeIterator.html
* `FusedIterator`
  * https://doc.rust-lang.org/std/iter/trait.FusedIterator.html

I don't think there's any reason to implement `DoubleEndedIterator` at this point.

File: zebra-state/src/service.rs
Comment (teor2345) [reply zebra-state/src/service.rs]: Oh, right, we need to access that enum variant when transitioning from non-finalized to finalized state.

I don't think this is a problem then.

File: zebra-state/src/service.rs
Comment (teor2345) [inline zebra-state/src/service.rs]: 👍

**Review by teor2345 — CHANGES_REQUESTED** [review]
Looks good, it's a bit simpler than I thought.

I have a few suggestions about side-chain and chain reorganisation semantics. I also found a bug where the iterator can't be used with a side-chain block hash.

Let's write tests for the side-chain (hash) and best chain (hash or height) cases.

File: zebra-state/src/service.rs
Comment (teor2345) [reply zebra-state/src/service.rs]: If that's what you decided, that's fine with me. I don't really need to repeat that discussion.

Comment (teor2345) [inline zebra-state/src/service.rs]: But it's still important that we document the precise semantics of the `chain` function.

I think they are:
```suggestion
    /// Return an iterator over the relevant chain of the block identified by
    /// `hash_or_height`. Heights iterate over the best chain. Hashes iterate
    /// over the chain containing the hash, which can be the best chain or a
    /// side chain. 
    #[allow(dead_code)]
    pub fn chain(&self, hash_or_height: HashOrHeight) -> Iter<'_> {
```

File: zebra-state/src/service.rs
Comment (teor2345) [reply zebra-state/src/service.rs]: Is this a standard property of rust iterators?
Or should we document it as part of the `chain` function docs?

File: zebra-state/src/service.rs
Comment (yaahc) [reply zebra-state/src/service.rs]: I started with that but that wouldn't apply here. The reason we access `NonFinalizedState` here isn't because it's newly setup, it's because the `hash` used in the last `next_non_finalized_state` is not a hash of non-finalized state, and so we need to use that same hash to lookup the height in the finalized state.

By accessing the HashOrHeight(hash) from the previous state at the start of `next_finalized_block` we make it so we only have to lookup the height for the hash once, and then we iterate by height which should be more sympathetic to the database design than continuing to iterate by previous hash.

We could change it so that if the `next_non_finalized_block` fails to find the block, it first attempts to resolve the height of the hash via the finalized state, but this would be doing work that wont be needed until the next iteration, which may never occur, so my preference would be to keep the current design.

File: zebra-state/src/service.rs
Comment (yaahc) [reply zebra-state/src/service.rs:368]: Since this is an entirely internal API I'd prefer to delay implementing these until we have a concrete use case.

File: zebra-state/src/service.rs
Comment (yaahc) [reply zebra-state/src/service.rs]: It's a property of borrows, this `Iter` type wraps a shared borrow of the `StateService`, which means it would be a compiler error to produce a mutable borrow to that same service, or any of its member data.

File: zebra-state/src/service.rs
Comment (teor2345) [reply zebra-state/src/service.rs:368]: `Iterator::size_hint` and `ExactSizeIterator` allow us to efficiently assert that the chain is at least 28 blocks long.

From the `FusedIterator` docs:
> Calling next on a fused iterator that has returned None once is guaranteed to return None again. This trait should be implemented by all iterators that behave this way because it allows optimizing Iterator::fuse.

There's no particular reason to implement `FusedIterator` right now, but it also doesn't cost us anything. `FusedIterator` has no methods - it's only used to make `Iterator::fuse` into a no-op.

`fuse` is used internally by `flatten` and possibly other iterator methods. So I don't think we'd necessarily notice when we started using `fuse`, and even if we did, we might not remember to implement `FusedIterator` on the `chain` iterator.

**Comment by yaahc** [comment]
@teor2345 I changed the API to sidestep the best chain caveats that were getting added. Since I expect this API will only ever be called by the block difficulty check and that check will have a block hash I went ahead and changed it to only take a `block::Hash` as the starting identifier for the relevant chain.

File: zebra-state/src/service.rs
Comment (yaahc) [reply zebra-state/src/service.rs:368]: > `Iterator::size_hint` and `ExactSizeIterator` allow us to efficiently assert that the chain is at least 28 blocks long.

This property isn't particularly useful though, because the caller `check::block_difficulty` or w/e its called will have access to the block itself, and will know how long its chain would be. Also, iirc we put in the RFC that we assume that there are at least 28 blocks of finalized state whenever we call `check::block_difficulty`, so we don't need to assert this in advanced, just panic if we find we have insufficient blocks while iterating.

File: zebra-state/src/service/memory_state/non_finalized_state.rs
Comment (teor2345) [inline zebra-state/src/service/memory_state/non_finalized_state.rs]:
```suggestion
    /// Returns the height of `hash` in any chain.
```

**Review by teor2345 — COMMENTED** [review]
The API looks good, let's

**Review by teor2345 — CHANGES_REQUESTED** [review]
The API looks good!

Let's write tests that:
* iterate from the non-finalized to the finalized state
* start iterating in the finalized state

**Comment by yaahc** [comment]
> 
> 
> The API looks good!
> 
> Let's write tests that:
> 
>     * iterate from the non-finalized to the finalized state
> 
>     * start iterating in the finalized state

Sounds good, I'm going to hold off doing this until I finish the other things I was already working on. Is the current content of this PR enough to unblock your work on the block difficulty check?

**Comment by yaahc** [comment]
actually, wait, I don't think we _can_ write those tests. Doing a test for iteration over the non_finalized state runs into the same issue as every other test, aka needing to first commit 500,000 blocks.

**Comment by teor2345** [comment]
> Since I expect this API will only ever be called by the block difficulty check and that check will have a block hash I went ahead and changed it to only take a `block::Hash` as the starting identifier for the relevant chain.

Just to clarify, I think that other contextual validation checks will use this API, but they'll all be supplying a hash for a specific chain. (Contextual checks will never iterate through the best chain by height.)

**Comment by teor2345** [comment]
> > The API looks good!
> > Let's write tests that:
> > ```
> > * iterate from the non-finalized to the finalized state
> > 
> > * start iterating in the finalized state
> > ```
> 
> Sounds good, I'm going to hold off doing this until I finish the other things I was already working on. Is the current content of this PR enough to unblock your work on the block difficulty check?

Yes, I'll summarise the interface in this PR in the contextual validation RFC.

> actually, wait, I don't think we _can_ write those tests. Doing a test for iteration over the non_finalized state runs into the same issue as every other test, aka needing to first commit 500,000 blocks.

You're right, and we can't even fake a non-finalized state using a private interface, because these tests are at the state service level.

But we can do this test:
> * start iterating in the finalized state

**Comment by hdevalence** [comment]
Do we want to get the full blocks here?

**Comment by teor2345** [comment]
> Do we want to get the full blocks here?

Good question! Are there any other contextual validation checks that will use the full blocks?

For example, the miner subsidy check needs the amounts from all the UTXOs that are spent in the block. But we'll probably do those checks using `AwaitUtxo` rather than `chain`.

Are there any other contextual validation checks which need transaction data? Or can we just get away with using headers?

**Comment by teor2345** [comment]
I need to merge this PR to start implementing the contextual difficulty RFC.

So I opened #1302 to follow up the tests and the potential refactor to use a `Header` iterator.

**Review by teor2345 — APPROVED** [review]
Moved remaining tasks to #1302.

**Comment by daira** [comment]
@teor2345 wrote:
> Are there any other contextual validation checks which need transaction data? Or can we just get away with using headers?

There's the "height in coinbase" check.

**Pull request merged**
